src/lib/es5.d.ts

```diff
@@ -1040,17 +1040,40 @@ interface JSON {
     /**
      * Converts a JavaScript value to a JavaScript Object Notation (JSON) string.
      * @param value A JavaScript value, usually an object or array, to be converted.
-     * @param replacer A function that transforms the results.
+     * @param replacer An array of strings and numbers that acts as a approved list for selecting the object properties that will be stringified.
      * @param space Adds indentation, white space, and line break characters to the return-value JSON text to make it easier to read.
      */
-    stringify(value: any, replacer?: (this: any, key: string, value: any) => any, space?: string | number): string;
+    stringify(
+        value: number | string | boolean | null | unknown[],
+        replacer?: (number | string)[] | null,
+        space?: string | number
+    ): string;
+    stringify(
+        value: undefined | Function | Symbol,
+        replacer?: (number | string)[] | null,
+        space?: any
+    ): undefined;
+    stringify(

```

Comment:
This can be combined with the signature on line 1046 -- it can just be `value: number | string | boolean | null | object` (`unknown[]` is a subtype of `object`).

Reply:
We need to separate the `object` input type out into a subsequent overload as otherwise it also matches a `Function` input value, i.e.

```ts
const a: string = JSON.stringify(() => null); // This is wrong
const b: undefined = JSON.stringify(() => null);
// Type 'string' is not assignable to type 'undefined'.
```

Separating out the overloads means that we don't need to make the type unnecessarily non-specific:

```ts
const c: string = JSON.stringify({...}); // 👍 
const d: string = JSON.stringify(new Object()); // 👍
const e: string = JSON.stringify(x as any);
// Type 'string | undefined' is not assignable to type 'string'. 👍
const f: string = JSON.stringify(() => null);
// Type 'undefined' is not assignable to type 'string'.
```

Reply:
Hmm, this is tricky. We don't want `JSON.stringify({ call: null, apply: null, bind: null})` to be treated as possibly returning `undefined`, but from TS's perspective, that's the same as what's happening when you pass a function. `stringify`'s behavior effectively breaks subtyping from our perspective since `func instanceof Object` is `true`.

The number of breaks in our own codebase alone implies this isn't going to be very palatable. Assuming the user tests show similar results, I'd prefer some set of overloads that result in `undefined` only being in the output type if it's also known to be in the input type.

Reply:
It is tricky indeed. 

`JSON.stringify({ call: null, apply: null, bind: null})` results in `string`, but maybe there's some more specific case that would falsely return `undefined`? The order of overloads does solve the problem of misrecognising a function as a valid input for stringify. However, I don't think functions passed to stringify is the biggest problem, so can make it return string for it (even though, of course, that would be false).

From the user tests I can see that there are some instances where `JSON.stringify(<any>)` results in `string | undefined`. It looks like in VSCode it may even be handled, but still `string` is not `string | undefined`. Would you suggest I pro-actively make pull-requests in these repos to remove these 'any's?

For TS to claim that `JSON.stringify(<any>)` always results in `string` is not right, and means all dependents need to handle this case at runtime.

Reply:
Well, you can't fix the entire world. We'll have to figure out what the right balance of correctness vs not breaking everyone is. I'll bring it up in our weekly design meeting.

Reply:
OK, so this seems better to me:

```
JSON.stringify(<undefined>) : undefined
JSON.stringify(<string | number | boolean | object>) : string
JSON.stringify(<any>) : any
```
I've descoped fixing Function input values, so they would still return `string`. To fix them we'd have to move its overload first, showing up first in documentation, which is not ideal.

Reply:
`stringify` is even harder to type correctly considering that any `object` (including functions) can have a `.toJSON()` that can in turn return something that may or may not be stringifiable:

![image](https://user-images.githubusercontent.com/3902892/82079192-4d570c80-9697-11ea-8d39-2983288a1d5f.png)

Reply:
@rbuckton Very interesting, forgot about toJSON. It seems like comprehensively typing stringify is not an option right now. It is tempting though. Falsely returning `string` for any input value is worse and not doing that would be a vast improvement.

Reply:
@jupiter Why do you think `any` could be suitable there? I didn't find any previous explanation.